### ISSUE TITLE
Checkpointed function does not preserve `requires_grad` if output is a dataclass.

### ISSUE THREAD

**Issue by AShedko** [pytorch/pytorch#124725]
### 🐛 Describe the bug

Checkpointed function does not preserve `requires_grad` if output is a dataclass.

Minimal repro:
```py
import torch
from dataclasses import dataclass
from dataclasses import astuple, dataclass
import torch.utils
import torch.utils.checkpoint
import lightning as L

@dataclass
class D:
    s: torch.tensor
    y: torch.tensor
    # easy unpacking
    def __iter__(self):
        return iter(astuple(self))


def foo_dataclass(x:torch.Tensor):
    return D(s=x.mean(), y=torch.ones_like(x))

def foo_tuple(x:torch.Tensor):
    return (x.mean(), torch.ones_like(x))
a = torch.randn(40)
a.requires_grad=True

print("Dataclass:")
r = torch.utils.checkpoint.checkpoint(foo_dataclass, a, use_reentrant=True)
print("Checkpointed: ", r.s.requires_grad)
r2 = foo_dataclass(a)
print("Non-Checkpointed: ", r2.s.requires_grad)

a = torch.randn(40)
a.requires_grad=True
print("Tuple:")
r = torch.utils.checkpoint.checkpoint(foo_tuple, a, use_reentrant=True)
print("Checkpointed: ", r[0].requires_grad)
r2 = foo_tuple(a)
print("Non-Checkpointed: ", r2[0].requires_grad)
```

output is:
```yaml
Dataclass:
  Checkpointed:  False
  Non-Checkpointed:  True
Tuple:
  Checkpointed:  True
  Non-Checkpointed:  True
```

`checkpoint` also decays a `namedtuple` into a `tuple`, but does, at least, preserve gradients.

### Versions

Couldn't import dot_parser, loading of dot files will not be possible.
Collecting environment information...
PyTorch version: 2.1.2+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.5.0-1ubuntu1~22.04) 9.5.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-25-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA RTX A5000
GPU 1: NVIDIA RTX A5000

Nvidia driver version: 535.171.04
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             32
On-line CPU(s) list:                0-31
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen Threadripper PRO 5955WX 16-Cores
CPU family:                         25
Model:                              8
Thread(s) per core:                 2
Core(s) per socket:                 16
Socket(s):                          1
Stepping:                           2
Frequency boost:                    enabled
CPU max MHz:                        7031.2500
CPU min MHz:                        1800.0000
BogoMIPS:                           7984.43
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin brs arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualisation:                     AMD-V
L1d cache:                          512 KiB (16 instances)
L1i cache:                          512 KiB (16 instances)
L2 cache:                           8 MiB (16 instances)
L3 cache:                           64 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-31
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Vulnerable: Safe RET, no microcode
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] pytorch-lightning==2.2.0.post0
[pip3] pytorch-memlab==0.3.0
[pip3] tinynumpy==1.2.1
[pip3] torch==2.1.2+cu118
[pip3] torchaudio==2.1.2+cu118
[pip3] torchmetrics==1.3.1
[pip3] torchvision==0.16.2+cu118
[pip3] torchviz==0.0.2
[pip3] triton==2.1.0
[conda] numpy                     1.26.4          py310hb13e2d6_0    conda-forge
[conda] pytorch-lightning         2.2.0.post0              pypi_0    pypi
[conda] pytorch-memlab            0.3.0                    pypi_0    pypi
[conda] tinynumpy                 1.2.1                    pypi_0    pypi
[conda] torch                     2.1.2+cu118              pypi_0    pypi
[conda] torchaudio                2.1.2+cu118              pypi_0    pypi
[conda] torchmetrics              1.3.1                    pypi_0    pypi
[conda] torchvision               0.16.2+cu118             pypi_0    pypi
[conda] torchviz                  0.0.2                    pypi_0    pypi
[conda] triton                    2.1.0                    pypi_0    pypi

cc @soulitzer

**Comment by soulitzer** [comment]
Thanks for the report. Unfortunately, this is expected behavior due to limitation of `use_reentrant=True` activation checkpoint. 

> The reentrant version does not consider tensors in nested structures (e.g., custom objects, lists, dicts, etc) as participating in autograd, while the non-reentrant version does.

(and this is somewhat hinted at the docs.)

The behavior is correct when using the recommended setting - `use_reentrant=False`, does that work for your case?

**Comment by soulitzer** [comment]
Closing this because it is expected behavior.
